### DISCUSSION TITLE
Extension Nodes: Why have three script nodes?

### DISCUSSION THREAD

**Discussion by slanden** [pdml-lang/pdml-lang.github.io#9] [Q&A] [answered]
There are three script nodes in PDML's Extension Nodes

- s:script
- s:exp
- s:def

After reading [the docs](https://pdml-lang.dev/docs/extensions/user_manual/index.html#script_nodes) My understanding is as follows:

An `s:script` node allows embedding script code, wrapped inside code fences.

An `s:exp` node is like an `s:script` node, but must be only one line.

An `s:def` node is like an `s:script` node that can only contain definitions.

The code inside all of these are parsed by the scripting language, which doesn't make such a split. To implement `s:def`, a PDML parser would need to have code that parses the scripting language to check for and prohibit functionality available in the scripting language.

What's the purpose of having this split?

P.S. I posted here in _basic-specification_ because I didn't see a separate discussion for Extensions.

**Comment by pml-lang** [comment]
> What's the purpose of having this split?

The three nodes have different semantics and are handled differently.

- `s:exp`: The text content of the node must be a valid Javascript expression (e.g. `[s:exp 1 + 1]`). The expression is evaluated (e.g. `2`), converted into a string (e.g. `"2"`), and the resulting string is then inserted into the document.

- `s:script`: The content must be a set of valid Javascript instructions (e.g. an `if then else` instruction, loops, etc.). The instructions are executed and the outcome depends on the instructions. For example, you could use this to execute an external program, and write the program's output into the PDML document.

- `s:def`: This node typically contains function and constant declarations that are later used in `s:exp` and `s:script` nodes. This code is _not_ executed immediately. It is executed when it is called from an `s:exp` or `s:script` node.

All three nodes are of type [raw_text](https://pdml-lang.dev/docs/extensions/reference_manual/index.html#raw_text) (note that the 'Text Block Syntax' is deprecated and will be removed in a future version). Hence, the following two syntaxes are valid for all three nodes:

    [s:{name} code]

    [s:{name}
        ~~~
        code
        ~~~
    ]

For examples of how to use script nodes please have a look at [PDML Scripting Examples](https://pdml-lang.dev/docs/extensions/scripting_examples/index.html).

**Reply by slanden** [reply]
Can you give an example of when you would _need_ to use an `s:exp` node because an `s:script` node wouldn't work?

I see the semantic difference between a single expression and a number of statements with multiple expressions, but I don't see why you wouldn't just use `s:script`. In both cases, it seems the `[s:{script, exp} ...]` node is evaluated and replaced by the output of the node. If there was no output, there would be... well, it would be replaced with nothing. Right?

For the `s:def` node, how would the parser know what definitions are available for `s:{script, exp}` nodes to use if they're not executed yet? The code in `s:def` needs to be _interpreted_ to register the definitions into the JavaScript context, making them usable to any code called later.

**Comment by pml-lang** [comment, marked as answer]
> Can you give an example of when you would need to use an s:exp node because an s:script node wouldn't work?

You _can_ use a `s:script` node to achieve the same as with an `s:exp` node, because script nodes allow you to explicitly insert text into the document. However, if you just want to insert the result of an expression into the document (happens often), then the `s:exp` is much more convenient.

For example if you want to compute `1 + 1` (or any other expression of arbitrary complexity) and insert the result into the document, then you can simply use an `s:exp` node:

    [s:exp 1 + 1]

You could achieve the same with a `s:script` node, but it's obviously more verbose:

    [s:script doc.insertText(String(1+1));]

An `s:exp` node can only contain _a single expression_ and the result of evaluating the expression is _implicitly_ inserted into the document.
A `s:script` node is more powerful because it can contain _one or more Javascript instructions_. But you need to _explicitly_ call `doc.insertText` if you want to insert text into the document.

> In both cases, it seems the [s:{script, exp} ...] node is evaluated and replaced by the output of the node.

No, this is only done for `s:exp` nodes.

> For the s:def node, how would the parser know what definitions are available for s:{script, exp} nodes to use if they're not executed yet?

The content of `s:def` nodes is stored internally in the Javascript engine (embedded in PMLC), and executed whenever an `s:exp` or `s:script` node calls a function, or uses a constant defined in an `s:def` node.

**Reply by slanden** [reply]
Thank you for taking the time to go more in depth on this topic.

TL;DR
I see essentially two use cases being split by three nodes, and think either `s:script` and `s:exp` or `s:script` and `s:def` *could* be consolidated. The current `s:def` is confusing because although it uses the same scripting language (JavaScript) as the other script nodes, you could have valid JavaScript that PDML is not allowing. Perhaps if a second type of script node is necessary, the deciding factor is not definitions vs statements/expressions, but rather _when_ the code is used. I.e. static vs dynamic, compile time vs run time, eager vs lazy.

---

> you need to explicitly call doc.insertText if you want to insert text into the document.

The implicit vs explicit insert behavior for `s:exp` and `s:script` respectively clears things up a bit. To draw similarities to HTML, `s:script` is similar to HTML's `<script>` node, while `s:exp` is similar to 3rd party "templating" libraries, e.g. Handlebars.js. "Templating" is the one thing all web frameworks add, so I appreciate this being built into PDML's scripting capabilities via `s:exp`.

My curiosity is from thinking more about the reason scripts are put in markup to begin with. When a script is embedded in the markup, the goal is either

- A) To insert the result into the page, whether it's an element or text
- B) To define things or declare dependencies to be used by other scripts and event listeners **later**

In the case of *A*, even if you run multiple statements and external programs, the end goal is to insert the result into the document. Even if the script is only run for its side effects, the script was run and the node was parsed and no longer needed. Implicitly inserting the result and replacing it with its output (which can be nothing) should be fine for `s:script` too, therefore `s:script` and `s:exp` could be consolidated.

It'd be nice to pre-compute as much as possible, but that's out of scope; If browsers could run PDML directly, they would assume the file is fully functional without needing a build step.

In the case of *B*, I can see the need for a place to put definitions or tell the browser about dependencies or library code, etc. Then again, an `s:def` node only exists in the markup to insert things into the JavaScript context. Once parsed, nothing else happens with it; it's "used up".

To add to that point.. To include a source file, the other source file would have to be inside an `s:def` node, inside a PDML file.

```
[u:ins_file path=path/to/script.pdml]
```

That inserts the `s:def` node(s) into the current file, but `s:def` nodes only exist to insert declarations/definitions into the scripting context. They invoke the scripting language to set the items in memory, but JavaScript allows this already. Using a regular script node could avoid that.

```
[s:script (src=path/to/script.js)]
```

To load the script lazily, you'd probably do something like

```
[s:script doc.insert('[s:script (src=path/to/script.js)]')]
```

from an event handler somewhere.

So, it's even possible that use case *B* could be accounted for with the same node as use case *A*, because even though nothing is rendered to the page the functions and variables were still defined. However, this would mean you couldn't see them in the DOM alongside the other nodes since the node itself would be replaced by its output (which is nothing). Then again, does that even matter..

**Reply by pml-lang** [reply]
> I see essentially two use cases being split by three nodes, and think either s:script and s:exp or s:script and s:def could be consolidated.

I am always in favor of simplyfying things as much as possible.

It would indeed be possible to consolidate `s:script` and `s:def` nodes, because Javascript allows to mix functions with statements that are not part of a function.
In fact, PDML _does_ already allow to mix statements and functions in a `s:script` node (as can be done in HTML). Example:

```
[s:script
    function add ( a, b ) {
        return a + b;
    }

    doc.insertText ( add ( 1, 1 ) );
]
```

Note that a function defined in a script node can (in PDML) only be used in the script node that contains the function. But that could be changed (i.e. the function could be used in other script nodes too). Hence, the `s:def` node could be removed.

It _might_ even be possible to remove the `s:exp` node too, by improving `s:script`. For example, we could say that if the last statement in a `s:script` node is an expression, then the result of evaluating the expression is inserted into the document, else nothing is inserted.

However, I think that we should not do these changes for the following reasons (sorted roughly in order of importance):

1

While reducing the number of nodes would be possible for script nodes that contain _Javascript_, these simplification would not work for other programming languages that PDML might support in the future. For example, it would be great if scripting nodes could be written in Java (or [other languages supported by GraalVM](https://www.graalvm.org/22.0/reference-manual/languages/)). In some languages (e.g. Java), it is not possible to mix functions with statements that are not part of a function/method contained in a class. Hence, it is necessary to make a distinction between `s:script` and `s:def` nodes. (Note: It would theoretically be possible for the PDML parser to parse the Java code and separate functions from statements, and handle them appropriately. But that would be very very challenging, error-prone, and slow down execution.)

Note: If you look at JSP, you can see that JSP makes the same conceptual separation as in PDML:

- [Scriptlet tag](https://www.w3schools.blog/jsp-scriptlet-tag)
- [Declaration tag](https://www.w3schools.blog/jsp-declaration-tag)
- [Expression tag](https://www.w3schools.blog/jsp-expression-tag)

Note: [EJS](https://ejs.co/#docs) (a Javascript text template engine) uses two tags: script tag (`<% %>`, called scriptlet), and exp tag (`<%= %>`). It doesn't need a `def` tag for the reason explained above.

Note: I am not a fan of Javascript (to put it mildly). The reason to support Javascript in PDML was that Javascript is very popular, has many ready-to-use libraries, and is currently the only language really well supported as a scripting language in GraalVM.

2

The three nodes have different semantics and rules, and are handled differently. Therefore it makes sense to keep them separated.

- content
    - exp: contains a single expression
    - script: contains a set of statements (like the body of a function)
    - def: contains a set of variables, constants, and/or functions that can be used in subsequent `exp` and `script` nodes. You can think of it as library code. 

- handling
    - exp: the expression is evaluated, converted into a string, and written into the document
    - script: the statements are executed
    - def: the code is saved for subsequent usage

- node inserts text into the document
    - exp: yes
    - script: maybe
    - def: no

- code is executed immediately
    - exp: yes
    - script: yes
    - def: no

- code is retained in memory
    - exp: no
    - script: no
    - def: yes

3

Distinguishing between `exp`, `script`, and `def` nodes makes the code more clear in some situations. For example, if you see a `def` node, you know immediately that the code is not executed immediately. The intent of the node is clear: provide shared code that will be used in subsequent `exp` and/or `script` nodes.

4

Distinguishing between `exp`, `script`, and `def` nodes allows PDML to make some optimizations. For example, only `def` nodes need to be kept in memory. Or, a preprocessor could read the whole document, and place all `s:def` nodes at the start of the document, so that `s:exp` and `s:script` nodes can use `s:def` nodes that are defined later in the document (e.g. at the end of the document).


> My curiosity is from thinking more about the reason scripts are put in markup to begin with. When a script is embedded in the markup, the goal is either
    A) To insert the result into the page, whether it's an element or text
    B) To define things or declare dependencies to be used by other scripts and event listeners later

A third use case is to produce a side effect or do anything else you can do with the scripting language (e.g. write to STDOUT, log to a file, display a GUI dialog, etc., as shown in [PDML Scripting Examples](https://pdml-lang.dev/docs/extensions/scripting_examples/index.html))

> To include a source file, the other source file would have to be inside an s:def node

Note: You can also use `u:ins_file` to import `s:exp` or `s:script` nodes.

**Reply by slanden** [reply]
> In some languages (e.g. Java), it is not possible to mix functions with statements that are not part of a function/method contained in a class.

This perspective helps a lot in understanding the reasoning. Having different semantics and handling, in my opinion, doesn't make much sense with scripting languages; a characteristic of them is the ability to have everything at the top level, e.g. calling statements outside of a "main" function. If compiled languages would be supported, drawing a semantic line seems more reasonable.

Definitions in a compiled language wouldn't be executed immediately, but a scripting language would have to interpret that code to make them available to other scripting nodes. That contradicts how `s:def` is meant to be handled.

> you can see that JSP makes the same conceptual separation as in PDML

Another piece of interesting history is that c/c++ has the concept of header files... Ah, in all seriousness though, when I was first reading about `s:def` I thought, "Oh no..." because it reminded me if that concept. I know it's not the same, but just the separation reminded me of all the header files I don't miss.

>  EJS (a Javascript text template engine) uses two tags: script tag (<% %>, called scriptlet), and exp tag (<%= %>)

I can imagine not liking this at all. `s:exp` and `s:def` is at least a clear separation. When the '=' in `<%= %>` is all that separates the different behaviors, I'd probably spend too much time debugging typos. If there must be two things, I think it's best not to be subtle.

> Distinguishing between exp, script, and def nodes makes the code more clear in some situations.

While I now understand the reason for `s:def`, I still think there's a possibility to combine `s:exp` and `s:script` even when the "script" is for a compiled language. However, I'd rather revisit the idea after exploring how this would function with real code using a compiled language.

Perhaps with `[s:script (lang=java) ... ]`
the `lang` attribute selects a compiler to use for that language. Such a handler could wrap the contents of the node inside a function with a generated unique function name. There would be a "main" function associated with the PDML source file that would call the other generated functions. Hm, maybe..

**Reply by pml-lang** [reply]
> Perhaps with [s:script (lang=java) ... ] the lang attribute selects a compiler to use for that language.

Yes, that's what might be done in the future (in addition to a global parameter to set the default value for the `lang` attribute).

FYI: [Java on Truffle](https://www.graalvm.org/22.0/reference-manual/java-on-truffle/) (currently an experimental feature) will probably enable us to easily support Java as a scripting language in PDML/PML. More info: [Java on Truffle — Going Fully Metacircular](https://medium.com/graalvm/java-on-truffle-going-fully-metacircular-215531e3f840), [Java on Truffle](https://www.graalvm.org/java-on-truffle/)
